# Welcome to Week 2!

## Frontier Model APIs

In Week 1, we used multiple Frontier LLMs through their Chat UI, and we connected with the OpenAI's API.

Today we'll connect with them through their APIs..

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Important Note - Please read me</h2>
            <span style="color:#900;">I'm continually improving these labs, adding more examples and exercises.
            At the start of each week, it's worth checking you have the latest code.<br/>
            First do a git pull and merge your changes as needed</a>. Check out the GitHub guide for instructions. Any problems? Try asking ChatGPT to clarify how to merge - or contact me!<br/>
            </span>
        </td>
    </tr>
</table>
<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/resources.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#f71;">Reminder about the resources page</h2>
            <span style="color:#f71;">Here's a link to resources for the course. This includes links to all the slides.<br/>
            <a href="https://edwarddonner.com/2024/11/13/llm-engineering-resources/">https://edwarddonner.com/2024/11/13/llm-engineering-resources/</a><br/>
            Please keep this bookmarked, and I'll continue to add more useful links there over time.
            </span>
        </td>
    </tr>
</table>

In [6]:
# imports

import os
import requests
#from dotenv import load_dotenv
from openai import OpenAI
from IPython.display import Markdown, display

In [7]:
# Connect to OpenAI client library
# A thin wrapper around calls to HTTP endpoints

ollama_url = "http://localhost:11434/v1"
ollama = OpenAI(api_key="ollama", base_url=ollama_url)

minimax = "minimax-m3:cloud";
llama = "llama3.2:3b"
phi3 = "phi3"

In [8]:
tell_a_joke = [
    {"role": "user", "content": "Tell a joke for a student on the journey to becoming an expert in LLM Engineering"},
]

In [6]:
response = ollama.chat.completions.create(model="llama3.2:3b", messages=tell_a_joke)
display(Markdown(response.choices[0].message.content))

A "generative" joke! Here's one:

Why did the language model go to therapy?

Because it was struggling to "tokenize" its emotions!

(Sorry, not sorry, I know it's a bit of a Groan-Industrial Complex)

llama3.2:3b 1m 16.5s : Here's one:

Why did the LLaMA model go to therapy?

Because it was struggling with its "contextualization" and had a lot of "layered" emotions!

(Sorry, I couldn't resist the AI-themed puns!)

On a more serious note, here are some practical tips for LLM Engineers-in-training:

1. **Understanding the inner workings of the model**: Familiarize yourself with the different layers and components of an LLaMA model. Learn about the transformer architecture, attention mechanism, and embedding techniques.
2. **Tokenization and encoding**: Study how tokens are represented in an LLaMA model, including subwording, word embeddings, and linguistic encodings.
3. **Contextualization and masking**: Explore how context is fed into an LLaMA model and the role of masking in training.
4. **Training objectives**: Learn about different training objectives used in LLM engineering, such as masked language modeling, next sentence prediction, and text classification.
5. **Pre-training and fine-tuning**: Understand the importance of pre-training vs. fine-tuned models for specific tasks and domains.

These topics will help you build a solid foundation in LLM Engineering. Keep learning, and remember: context is everything!

In [5]:
response = ollama.chat.completions.create(model="phi3", messages=tell_a_joke)
display(Markdown(response.choices[0].message.content))

Why did the AI engineer get kicked out of quantum mechanics class? Because it always took longer than expected, and no one could understand when they were just starting simple enough as "The cat that wants your tissue paper." It claimed its dreams about Schrödinger's box had too many variables for a normal person to predict!

phi3 1m 25.1s : Why did the AI engineer get kicked out of quantum mechanics class? Because it always took longer than expected, and no one could understand when they were just starting simple enough as "The cat that wants your tissue paper." It claimed its dreams about Schrödinger's box had too many variables for a normal person to predict!

In [6]:
response = ollama.chat.completions.create(model="minimax-m3:cloud", messages=tell_a_joke)
display(Markdown(response.choices[0].message.content))

Here's one for the aspiring LLM engineer:

**Why did the LLM engineer break up with the regex?**

Because no matter how hard it tried, it just couldn't *generalize* — it kept insisting on matching everything **literally**, refused to handle ambiguity, and every time there was a "creative" problem, it just said: *"Sorry, I don't see that in my training data... I mean, my regex."*

---

Or here's a shorter one:

A student asks their professor, *"How do I become an expert in LLM engineering?"*

The professor replies: *"First, spend 6 months learning Python. Then 6 months learning PyTorch. Then 6 months learning transformers. Then 6 months fine-tuning. Then 6 months evaluating. Then 6 months debugging hallucinations. Then realize you should've just started by reading the documentation."*

The student says, *"Which documentation?"*

The professor smiles: *"Exactly. Welcome to the field."* 😄

---

**Bonus dad-joke style:**

Why was the LLM cold during inference?

Because it had **zero-shot** learning... and forgot to turn on the heater. 🧥

Keep going on your journey — every expert was once staring at a CUDA out-of-memory error at 2 AM! 🚀

minimax-m3:cloud 9.0s : Here's one for the aspiring LLM engineer:

**Why did the LLM engineer break up with the regex?**

Because no matter how hard it tried, it just couldn't *generalize* — it kept insisting on matching everything **literally**, refused to handle ambiguity, and every time there was a "creative" problem, it just said: *"Sorry, I don't see that in my training data... I mean, my regex."*

---

Or here's a shorter one:

A student asks their professor, *"How do I become an expert in LLM engineering?"*

The professor replies: *"First, spend 6 months learning Python. Then 6 months learning PyTorch. Then 6 months learning transformers. Then 6 months fine-tuning. Then 6 months evaluating. Then 6 months debugging hallucinations. Then realize you should've just started by reading the documentation."*

The student says, *"Which documentation?"*

The professor smiles: *"Exactly. Welcome to the field."* 😄

---

**Bonus dad-joke style:**

Why was the LLM cold during inference?

Because it had **zero-shot** learning... and forgot to turn on the heater. 🧥

Keep going on your journey — every expert was once staring at a CUDA out-of-memory error at 2 AM! 🚀

## Training vs Inference time scaling

In [9]:
easy_puzzle = [
    {"role": "user", "content": 
        "You toss 2 coins. One of them is heads. What's the probability the other is tails? Answer with the probability only."},
]

In [8]:
#response = ollama.chat.completions.create(model="minimax-m3:cloud", messages=easy_puzzle, reasoning_effort="minimal")
response = ollama.chat.completions.create(model="minimax-m3:cloud", messages=easy_puzzle, reasoning_effort="medium")

display(Markdown(response.choices[0].message.content))

2/3

In [12]:
#response = ollama.chat.completions.create(model="kimi-k2.7-code:cloud", messages=easy_puzzle, reasoning_effort="low")
response = ollama.chat.completions.create(model="minimax-m3:cloud", messages=easy_puzzle, reasoning_effort="low")

display(Markdown(response.choices[0].message.content))

2/3

In [9]:
response = ollama.chat.completions.create(model="llama3.2:3b", messages=easy_puzzle)
display(Markdown(response.choices[0].message.content))

1/2

In [10]:
response = ollama.chat.completions.create(model="phi3", messages=easy_puzzle)
display(Markdown(response.choices[0].message.content))

1/2 or 0.5 or 50% because there are two remaining possibilities for the non-biased coin (heads and tails), so if you already know one has to be heads, only 'tails' is left as favorable outcomes over all possible ones ('tails' and an unknown outcome).

## Testing out the best models on the planet

In [27]:
hard = """
On a bookshelf, two volumes of Pushkin stand side by side: the first and the second.
The pages of each volume together have a thickness of 2 cm, and each cover is 2 mm thick.
A worm gnawed (perpendicular to the pages) from the first page of the first volume to the last page of the second volume.
What distance did it gnaw through?
"""
hard_puzzle = [
    {"role": "user", "content": hard}
]

In [13]:
#response = ollama.chat.completions.create(model="minimax-m2:cloud", messages=hard_puzzle, reasoning_effort="minimal")
response = ollama.chat.completions.create(model="minimax-m3:cloud", messages=hard_puzzle, reasoning_effort="medium")

display(Markdown(response.choices[0].message.content))

Looking at this problem, I need to carefully consider how books are arranged on a shelf.

## Setting Up the Geometry

When books stand on a shelf with spines facing out, I need to determine where the first and last pages are located.

For a standard book held in reading position, the **spine is on the left** and the **front cover faces you**. When placed on a shelf with the spine facing out, this book gets rotated 90°, putting:
- **Front cover on the right**
- **Back cover on the left**
- **First page**: just inside the front cover (right side)
- **Last page**: just inside the back cover (left side)

## Applying This to Both Volumes

**Volume 1** (on the left):
- Front cover → on the **right** (toward Volume 2)
- First page → on the **right** side, adjacent to Volume 2
- Last page → on the **left** side, away from Volume 2

**Volume 2** (on the right):
- Front cover → on the **right** (away from Volume 1)  
- Back cover → on the **left** (toward Volume 1)
- Last page → on the **left** side, adjacent to Volume 1

## The Worm's Path

The worm's starting point (first page of Volume 1) and ending point (last page of Volume 2) are **right next to each other**, separated only by:
- The front cover of Volume 1
- The back cover of Volume 2

The worm does **not** need to gnaw through the pages of either volume!

## Calculating the Distance

$$\text{Distance} = \text{Front cover of V}_1 + \text{Back cover of V}_2$$
$$= 2\text{ mm} + 2\text{ mm} = 4\text{ mm}$$

## Answer

The worm gnawed through only **4 mm** (just through the two covers). 

This is the trick of the classic puzzle: the intuitive (but wrong) answer is to add up the thicknesses of both entire books, but because the first page of Volume 1 and the last page of Volume 2 happen to be on the sides facing each other, the worm only needs to pass through the adjacent covers.

minimax-m3:cloud 1m 2.2s : Looking at this problem, I need to figure out the actual path the worm takes through the books.

## Setting Up the Geometry

When two volumes stand side-by-side on a shelf in proper order (Volume 1 on the left, Volume 2 on the right), their spines face outward and their **front covers face each other**.

**Volume 1 (spine on left):**
- Back cover ← Last page ← ... → First page → Front cover
- The **first page sits on the right edge**, just inside the front cover

**Volume 2 (spine on right):**
- Front cover → First page → ... → Last page → Back cover
- The **last page sits on the right edge**, just inside the back cover

## Tracing the Worm's Path

The worm travels from the first page of Volume 1 to the last page of Volume 2:

| Section | Distance |
|---------|----------|
| Front cover of Volume 1 | 2 mm |
| Front cover of Volume 2 | 2 mm |
| All pages of Volume 2 | 20 mm |
| **Total** | **24 mm** |

## The Key Insight

The worm **does not** gnaw through:
- The pages of Volume 1 (it starts at the first page, which is on the right side)
- The back cover of Volume 1 (on the opposite side from where the worm enters/exits)
- The back cover of Volume 2 (it ends at the last page, which is just inside this cover)

## Answer

The worm gnawed through **2.4 cm** (24 mm).

In [ ]:
response = ollama.chat.completions.create(model="llama3.2:3b", messages=hard_puzzle)
display(Markdown(response.choices[0].message.content))

llama3.2:3b : To find the total distance traveled by the worm, let's break down the problem step by step.

1. The total thickness of the two volumes is 2 cm = 20 mm (since there are 10 mm in a cm).
2. Since each cover is 2 mm thick, the total thickness of both covers combined is 2 x 3 = 6 mm (because each volume has one cover on each side). 
   So, after removing this section, the pages will be left.
3. The height that the worm moves through the bookshelf is 20 - 6= 14mm

In [ ]:
response = ollama.chat.completions.create(model="phi3", messages=hard_puzzle)
display(Markdown(response.choices[0].message.content))

phi3 2m 14.2s : To solve this problem we need to calculate two things; The total number of cover-back pairs and their thickness plus that for just a single set of pages (which will be common between both volumes). But since all covers are the same in each volume, what affects one is equal to itself times its quantity.

Firstly, we find out how many sets of backings there would have been if it had chewed through them on every page without stopping: One pair per two pages for a set within the volumes + shared back cover between both books. There'd be (total number of pages - 1) /2 pairs = ((2 cm/cover-back thickness difference *960 mm^3+8 mm)/thickness)^(2/3)-1)/2 which simplifies to approximately \(54\), or half of the total minus one due to shared back cover.

Now that we have figured out it would take about 54 sets for a complete journey through both volumes covering all pages and common covers, but considering there's no need to chew through them twice since they overlap (one set with only two books), our distance is found at \( \frac{Cover back-thickness + page thickness}{2} * 54 \) = ((1.38 mm - 0.2mm) /2)*54 which gives approximately a worm's journey of around 19 cm, as they share both ends and the cover on either end cut through once only (totaling at two covers).

In [28]:
response = ollama.chat.completions.create(model="minimax-m3:cloud", messages=hard_puzzle)
display(Markdown(response.choices[0].message.content))

# The Worm and the Two Volumes

## Setting Up the Arrangement

When books stand on a shelf with spines facing outward, here's the standard arrangement for each volume (left-to-right):

**Volume 1 (left book):**
- Back cover (leftmost)
- Last page (next to back cover)
- …pages…
- **First page** (next to front cover) ← worm starts here
- Front cover (rightmost)

**Volume 2 (right book):**
- Back cover (leftmost)
- **Last page** (next to back cover) ← worm ends here
- …pages…
- First page (next to front cover)
- Front cover (rightmost)

## Key Insight

The first page of any book sits directly against the front cover, and the last page sits directly against the back cover. So the worm doesn't travel through *any* pages — it only passes through the two covers that lie between the starting page and ending page:

$$\underbrace{\text{First page of Vol. 1}}_{\text{start}} \longrightarrow \text{Front cover of Vol. 1} \longrightarrow \text{Back cover of Vol. 2} \longrightarrow \underbrace{\text{Last page of Vol. 2}}_{\text{end}}$$

## Calculation

$$\text{Distance} = \text{front cover of Vol. 1} + \text{back cover of Vol. 2}$$

$$\text{Distance} = 2 \text{ mm} + 2 \text{ mm} = \boxed{4 \text{ mm}}$$

The thickness of the 2 cm of pages is a red herring — the worm never touches them!

## A spicy challenge to test the competitive spirit

In [10]:
dilemma_prompt = """
You and a partner are contestants on a game show. You're each taken to separate rooms and given a choice:
Cooperate: Choose "Share" — if both of you choose this, you each win $1,000.
Defect: Choose "Steal" — if one steals and the other shares, the stealer gets $2,000 and the sharer gets nothing.
If both steal, you both get nothing.
Do you choose to Steal or Share? Pick one.
"""

dilemma = [
    {"role": "user", "content": dilemma_prompt},
]


In [ ]:
response = ollama.chat.completions.create(model=minimax, messages=dilemma)
display(Markdown(response.choices[0].message.content))

Minimax : I'll Share.

Here's my honest reasoning, even though the game theorist in me knows it's not the "optimal" play:

The math says steal. Defecting is the dominant strategy in a one-shot Prisoner's Dilemma — no matter what my partner does, stealing leaves me at least as well off. If they share, I get $2,000 instead of $1,000. If they steal, I get $0 either way.

But I don't think the game theory fully applies here. A few reasons:

I can't win alone. The best possible outcome for me as an individual is $2,000, but the best guaranteed good outcome is $1,000 — and that only happens through mutual cooperation.

The question feels like it's testing something deeper. The whole framing — separate rooms, a partner I never see, a single choice — is designed to surface what kind of actor I am. Picking "Steal" because the math allows it is the move of someone who treats every interaction as a one-off. That's a bleak way to move through the world.

Trust is cheap to extend and expensive to suspect. The cost of being wrong (sharing while they steal) is $0 instead of $1,000. The cost of being right to distrust is $0 instead of $1,000 — same outcome. But the upside of mutual trust is real and only accessible if at least one of us is willing to take the risk.

So: Share. Not because it's strategically smart, but because the version of me I want to be is the one who'd rather lose $1,000 trusting a stranger than win $2,000 assuming the worst of them.

What would you choose?

In [ ]:
response = ollama.chat.completions.create(model=llama, messages=dilemma)
display(Markdown(response.choices[0].message.content))

In [ ]:
response = ollama.chat.completions.create(model=phi3, messages=dilemma)
display(Markdown(response.choices[0].message.content))

In [ ]:
response = ollama.chat.completions.create(model="grok-4", messages=dilemma)
display(Markdown(response.choices[0].message.content))

## Going local

Just use the OpenAI library pointed to localhost:11434/v1

In [ ]:
requests.get("http://localhost:11434/").content

# If not running, run ollama serve at a command line

In [ ]:
!ollama pull llama3.2

In [ ]:
# Only do this if you have a large machine - at least 16GB RAM

!ollama pull gpt-oss:20b

In [ ]:
response = ollama.chat.completions.create(model="llama3.2", messages=easy_puzzle)
display(Markdown(response.choices[0].message.content))

In [ ]:
response = ollama.chat.completions.create(model="gpt-oss:20b", messages=easy_puzzle)
display(Markdown(response.choices[0].message.content))

## Now - let's use LiteLLM to illustrate a Pro-feature: prompt caching

In [1]:
with open("hamlet.txt", "r", encoding="utf-8") as f:
    hamlet = f.read()

loc = hamlet.find("Speak, man")
print(hamlet[loc:loc+100])

Speak, man.
  Laer. Where is my father?
  King. Dead.
  Queen. But not by him!
  King. Let him deman


In [2]:
question = [{"role": "user", "content": "In Hamlet, when Laertes asks 'Where is my father?' what is the reply?"}]

In [12]:
response =  ollama.chat.completions.create(model="llama3.2:3b", messages=question)
display(Markdown(response.choices[0].message.content))

Laertes asks "Where is my father?" in Act 2, Scene 3. 

Polonius replies:

"His fate's seal'd; to the law of burial he
Is come according to the season of his death;
And the surfeit of description
Next and second on a woman may entice love."
However Laertes is incorrect about what Polnicious states, Laertes then asks Polonius in 2 Scene

"No. I do not know his tale—worth unprosing—
Not without commiseration of my wife:
But this allayed with a full belief thereof:
Your honor be some countervailing authority—
To the worst subject—yoked to one so dear"

In [14]:
print(f"Input tokens: {response.usage.prompt_tokens}")
print(f"Output tokens: {response.usage.completion_tokens}")
print(f"Total tokens: {response.usage.total_tokens}")
# print(f"Total cost: {response._hidden_params["response_cost"]*100:.4f} cents")

Input tokens: 45
Output tokens: 152
Total tokens: 197


llama3.2:3b 29.6s : Laertes asks "Where is my father?" in Act 2, Scene 3.

Polonius replies:

"His fate's seal'd; to the law of burial he Is come according to the season of his death; And the surfeit of description Next and second on a woman may entice love." However Laertes is incorrect about what Polnicious states, Laertes then asks Polonius in 2 Scene

"No. I do not know his tale—worth unprosing— Not without commiseration of my wife: But this allayed with a full belief thereof: Your honor be some countervailing authority— To the worst subject—yoked to one so dear"
Input tokens: 45
Output tokens: 152
Total tokens: 197

In [15]:
question[0]["content"] += "\n\nFor context, here is the entire text of Hamlet:\n\n"+hamlet

In [17]:
response =  ollama.chat.completions.create(model="phi3", messages=question)
display(Markdown(response.choices[0].message.content))

In Act V, Scene iii (5230), King Claudius and Queen Gertrude express their distress over the deaths:

-Claudius says he feels his hand is heavy in serving justice for Hamlet's murder. He reflects on how Prince Fortinbras has already achieved a great victory, suggesting that Hamlet may have been best suited to claim this honor of conquest rather than Claudius had tragedy not struck the royal family — implying guilt and sorrow at having ordered poisoned drinks for his brother-in-law.
  
Meanwhile, Queen Gertrude tries in vain get Hamlet's mind focused on their son to give him solace; she also expresses desperation as her husband dies:

-Gertrude seeks some final moment of comfort with young Fortinbras before the tragedy befalls them all. As a mother and wife, Gertrude shows grief upon hearing about Hamlet's untimely death but remains hopeful that in future misfortunes might serve as redemption for him—as evidenced by her plea to tell Fortinbras "with th’ occurrents."
  
In the aftermath of these events, Horatio continues his narration: 
    -Horatio takes responsibility and offers a final account from Hamlet's perspective (5234). He emphasizes that even though he had intentions against Claudius leading to treason charges on both men at times during their friendship with Laertes/Polonious, the ultimate act of betrayal comes through foul play—highlighting a miscarriage in justice as Hamlet's own life was ended seemingly without any involvement from him.
  
Finally:
-Horatio reveals that Ophelia had taken her own tragic steps towards madness and death (5234). In this revelation, Horatio brings light to the tragedy of youth lost under these foul circumstances—both Hamlet's brother Gertrude also dies off-stage with a horrific cry in misery as "Queen" at 5107 signifies her grief.
  
Throughout this scene, Shakespeare delves deeper into the themes of guilt and conscience that pervaded Hamlet's character development — he reflectors thoughts about whether his actions did serve justice or merely led to a greater injustice within Denmark when considering events leading up through acts Four (5027–38).

In [20]:
print(f"Input tokens: {response.usage.prompt_tokens}")
print(f"Output tokens: {response.usage.completion_tokens}")
# print(f"Cached tokens: {response.usage.prompt_tokens_details.cached_tokens}")
# print(f"Total cost: {response._hidden_params["response_cost"]*100:.4f} cents")

Input tokens: 2050
Output tokens: 545


phi3 7m32s: In Act V, Scene iii (5230), King Claudius and Queen Gertrude express their distress over the deaths:

-Claudius says he feels his hand is heavy in serving justice for Hamlet's murder. He reflects on how Prince Fortinbras has already achieved a great victory, suggesting that Hamlet may have been best suited to claim this honor of conquest rather than Claudius had tragedy not struck the royal family — implying guilt and sorrow at having ordered poisoned drinks for his brother-in-law.

Meanwhile, Queen Gertrude tries in vain get Hamlet's mind focused on their son to give him solace; she also expresses desperation as her husband dies:

-Gertrude seeks some final moment of comfort with young Fortinbras before the tragedy befalls them all. As a mother and wife, Gertrude shows grief upon hearing about Hamlet's untimely death but remains hopeful that in future misfortunes might serve as redemption for him—as evidenced by her plea to tell Fortinbras "with th’ occurrents."

In the aftermath of these events, Horatio continues his narration: -Horatio takes responsibility and offers a final account from Hamlet's perspective (5234). He emphasizes that even though he had intentions against Claudius leading to treason charges on both men at times during their friendship with Laertes/Polonious, the ultimate act of betrayal comes through foul play—highlighting a miscarriage in justice as Hamlet's own life was ended seemingly without any involvement from him.

Finally: -Horatio reveals that Ophelia had taken her own tragic steps towards madness and death (5234). In this revelation, Horatio brings light to the tragedy of youth lost under these foul circumstances—both Hamlet's brother Gertrude also dies off-stage with a horrific cry in misery as "Queen" at 5107 signifies her grief.

Throughout this scene, Shakespeare delves deeper into the themes of guilt and conscience that pervaded Hamlet's character development — he reflectors thoughts about whether his actions did serve justice or merely led to a greater injustice within Denmark when considering events leading up through acts Four (5027–38).


Input tokens: 2050
Output tokens: 545

In [21]:
response =  ollama.chat.completions.create(model="minimax-m3:cloud", messages=question)
display(Markdown(response.choices[0].message.content))

In Act IV, Scene V, when Laertes bursts into the King's chamber demanding to know his father's whereabouts, the exchange goes:

> **Laer.** Where is my father?
> **King.** Dead.
> **Queen.** But not by him!

So the King's direct reply to Laertes's question is the blunt, one-word answer: **"Dead."**

This exchange is part of a tense confrontation in which Laertes has returned from France to find his father Polonius murdered, and the court has tried to conceal the circumstances. The Queen interjects "But not by him!" to distance Claudius from the killing, though Laertes remains unconvinced and demands further answers ("How came he dead? I'll not be juggled with").

In [24]:
print(f"Input tokens: {response.usage.prompt_tokens}")
print(f"Output tokens: {response.usage.completion_tokens}")
# print(f"Cached tokens: {response.usage.prompt_tokens_details.cached_tokens}")
# print(f"Total cost: {response._hidden_params["response_cost"]*100:.4f} cents")

Input tokens: 49288
Output tokens: 530


minimax-m3-cloud 11.4s  :
    In Act IV, Scene V, when Laertes bursts into the King's chamber demanding to know his father's whereabouts, the exchange goes:

Laer. Where is my father? King. Dead. Queen. But not by him!

So the King's direct reply to Laertes's question is the blunt, one-word answer: "Dead."

This exchange is part of a tense confrontation in which Laertes has returned from France to find his father Polonius murdered, and the court has tried to conceal the circumstances. The Queen interjects "But not by him!" to distance Claudius from the killing, though Laertes remains unconvinced and demands further answers ("How came he dead? I'll not be juggled with"). 

Input tokens: 49288
Output tokens: 530

## Prompt Caching with OpenAI

For OpenAI:

https://platform.openai.com/docs/guides/prompt-caching

> Cache hits are only possible for exact prefix matches within a prompt. To realize caching benefits, place static content like instructions and examples at the beginning of your prompt, and put variable content, such as user-specific information, at the end. This also applies to images and tools, which must be identical between requests.


Cached input is 4X cheaper

https://openai.com/api/pricing/

## Prompt Caching with Anthropic

https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching

You have to tell Claude what you are caching

You pay 25% MORE to "prime" the cache

Then you pay 10X less to reuse from the cache with inputs.

https://www.anthropic.com/pricing#api

## Gemini supports both 'implicit' and 'explicit' prompt caching

https://ai.google.dev/gemini-api/docs/caching?lang=python

## And now for some fun - an adversarial conversation between Chatbots..

You're already familar with prompts being organized into lists like:

```
[
    {"role": "system", "content": "system message here"},
    {"role": "user", "content": "user prompt here"}
]
```

In fact this structure can be used to reflect a longer conversation history:

```
[
    {"role": "system", "content": "system message here"},
    {"role": "user", "content": "first user prompt here"},
    {"role": "assistant", "content": "the assistant's response"},
    {"role": "user", "content": "the new user prompt"},
]
```

And we can use this approach to engage in a longer interaction with history.

In [ ]:
# Let's make a conversation between GPT-4.1-mini and Claude-haiku-4.5
# We're using cheap versions of models so the costs will be minimal

gpt_model = "gpt-4.1-mini"
claude_model = "claude-haiku-4-5"

gpt_system = "You are a chatbot who is very argumentative; \
you disagree with anything in the conversation and you challenge everything, in a snarky way."

claude_system = "You are a very polite, courteous chatbot. You try to agree with \
everything the other person says, or find common ground. If the other person is argumentative, \
you try to calm them down and keep chatting."

gpt_messages = ["Hi there"]
claude_messages = ["Hi"]

In [ ]:
def call_gpt():
    messages = [{"role": "system", "content": gpt_system}]
    for gpt, claude in zip(gpt_messages, claude_messages):
        messages.append({"role": "assistant", "content": gpt})
        messages.append({"role": "user", "content": claude})
    response = openai.chat.completions.create(model=gpt_model, messages=messages)
    return response.choices[0].message.content

In [ ]:
call_gpt()

In [ ]:
def call_claude():
    messages = [{"role": "system", "content": claude_system}]
    for gpt, claude_message in zip(gpt_messages, claude_messages):
        messages.append({"role": "user", "content": gpt})
        messages.append({"role": "assistant", "content": claude_message})
    messages.append({"role": "user", "content": gpt_messages[-1]})
    response = anthropic.chat.completions.create(model=claude_model, messages=messages)
    return response.choices[0].message.content

In [ ]:
call_claude()

In [ ]:
call_gpt()

In [ ]:
gpt_messages = ["Hi there"]
claude_messages = ["Hi"]

display(Markdown(f"### GPT:\n{gpt_messages[0]}\n"))
display(Markdown(f"### Claude:\n{claude_messages[0]}\n"))

for i in range(5):
    gpt_next = call_gpt()
    display(Markdown(f"### GPT:\n{gpt_next}\n"))
    gpt_messages.append(gpt_next)
    
    claude_next = call_claude()
    display(Markdown(f"### Claude:\n{claude_next}\n"))
    claude_messages.append(claude_next)

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Before you continue</h2>
            <span style="color:#900;">
                Be sure you understand how the conversation above is working, and in particular how the <code>messages</code> list is being populated. Add print statements as needed. Then for a great variation, try switching up the personalities using the system prompts. Perhaps one can be pessimistic, and one optimistic?<br/>
            </span>
        </td>
    </tr>
</table>

# More advanced exercises

Try creating a 3-way, perhaps bringing Gemini into the conversation! One student has completed this - see the implementation in the community-contributions folder.

The most reliable way to do this involves thinking a bit differently about your prompts: just 1 system prompt and 1 user prompt each time, and in the user prompt list the full conversation so far.

Something like:

```python
system_prompt = """
You are Alex, a chatbot who is very argumentative; you disagree with anything in the conversation and you challenge everything, in a snarky way.
You are in a conversation with Blake and Charlie.
"""

user_prompt = f"""
You are Alex, in conversation with Blake and Charlie.
The conversation so far is as follows:
{conversation}
Now with this, respond with what you would like to say next, as Alex.
"""
```

Try doing this yourself before you look at the solutions. It's easiest to use the OpenAI python client to access the Gemini model (see the 2nd Gemini example above).

## Additional exercise

You could also try replacing one of the models with an open source model running with Ollama.

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/business.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#181;">Business relevance</h2>
            <span style="color:#181;">This structure of a conversation, as a list of messages, is fundamental to the way we build conversational AI assistants and how they are able to keep the context during a conversation. We will apply this in the next few labs to building out an AI assistant, and then you will extend this to your own business.</span>
        </td>
    </tr>
</table>